rules/src/main/scala/com/mobimeo/gtfs/rules/Engine.scala

```diff
@@ -0,0 +1,190 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mobimeo.gtfs.rules
+
+import cats.data.NonEmptyList
+import cats.effect.Sync
+import cats.syntax.all._
+import com.mobimeo.gtfs.file.GtfsFile
+import fs2._
+import fs2.data.csv._
+import org.typelevel.log4cats.StructuredLogger
+import cats.data.OptionT
+
+/** An engine is used to apply set of processing rules to GTFS files. It provides a common environment composed of
+  * functions that can be used in the rules.
+  *
+  * An engine can be reused for several set of rules and GTFS files using the same set of environment.
+  */
+class Engine[F[_]](interpreter: Interpreter[F])(implicit
+    F: Sync[F],
+    logger: StructuredLogger[F]
+) {
+
+  /** Applies the rule sets to the `from` file, writing results in `to`.
+    *
+    * Rule sets are fused to avoid going over the same file several times when several rule sets apply to the same file.
+    * The order in which the files are processed is not specified (and can even be done in parallel) but the different
+    * rule sets operating on a single file are ensured to be run in the same order as provided.
+    */
+  def process(rulesets: List[RuleSet[F]], from: GtfsFile[F], to: GtfsFile[F], ctx: Ctx = Ctx.empty): F[Unit] = {
+    // fusing rule sets allows to only go through the entities once
+    // even if several rulesets are defined for a given file
+    // this is possible because each ruleset consider the rows in isolation
+    // so as long as the order is kept the semantic stays the same
+    val fusedRulesets = rulesets.groupMap(_.file)(rs => (rs.rules, rs.additions)).toList
+    fusedRulesets.traverse_ { case (file, rulessadditions) =>
+      val (ruless, additionss) = rulessadditions.unzip
+      logger.info(s"Apply rules to $file for feed from ${from.file} to ${to.file}") >>
+        from.read
+          .rawFile(file)
+          .pull
+          .peek1
+          .flatMap {
+            case Some((CsvRow(_, headers), s)) =>
+              (s.through(processRules(ruless, ctx)) ++ Stream
+                .emits(additionss.flatten)
+                .evalMap(CsvRow(_, headers).liftTo[F])).pull.echo
+            case None => Pull.done
+          }
+          .stream
+          .through(to.write.rawFile(file))
+          .compile
+          .drain
+    }
+  }
+
+  private def processRules(ruless: List[List[Rule[F]]], ctx: Ctx): Pipe[F, CsvRow[String], CsvRow[String]] =
+    _.evalMap(processRow(ruless, ctx, _)).unNone
+
+  private def processRow(ruless: List[List[Rule[F]]], ctx: Ctx, row: CsvRow[String]): F[Option[CsvRow[String]]] =
+    F.tailRecM((row, ruless)) {
```

Comment:
Feels a bit like a `foldLeftM` to me, but hard to check without trying, so might be nonsense. `tailRecM` is not a bad choice anyway.

Reply:
`foldLeftM` doesn't work in this case as there is a non-failing shortcut (in case row is deleted, we don't process further). `foldLeftM` only stops on failure.